torch/fx/experimental/proxy_tensor.py

```diff
@@ -106,23 +105,33 @@ def set_meta(proxy, val):
 
 
 def track_tensor(tensor, proxy, *, constant, tracer):
+    def try_set_proxy_slot(outer_s, proxy_callable, *args):
+        assert callable(proxy_callable)
+        if isinstance(outer_s, SymInt):
+            inner_s = outer_s.get_pyobj()
+            assert isinstance(inner_s, PySymInt)
+            proxy = None
+
+            def thunk():
+                nonlocal proxy
+                if proxy is None:
+                    proxy = proxy_callable(inner_s, *args)
+                return proxy
+            set_proxy_slot(inner_s, tracer, thunk)
+
     # The basic idea is that we need to associate each tensor/SymInt
     # with a Proxy.  How do we setup this association?  We just store
     # the proxy on the proxy slot of the object, keyed on the tracer
     # (so that if we have multiple tracers at the same time, they
     # don't clobber each other.)
     for i, s in enumerate(tensor.shape):
-        if isinstance(s, SymInt):
-            inner_s = s.get_pyobj()
-            assert isinstance(inner_s, PySymInt)
-            # TODO: improve naming
-            # TODO: lazily insert this into the graph only on first
-            # use?  Maybe complicated and DCE is a better idea
-            s_proxy = torch.ops.aten.sym_size(proxy, i)
-            set_meta(s_proxy, inner_s)
-            set_proxy_slot(inner_s, tracer, s_proxy)
-
-        # TODO: also do stride/numel
+        try_set_proxy_slot(s, lambda x, i: set_meta(torch.ops.aten.sym_size(proxy, i), x), i)
```

Comment:
i think i finally get this, but it would help to comment.

you're trying to defer the sym_size call since it triggers ProxySymDispatchMode tracing into the fx graph which is expensive.  but the result is you store a callable in the slot instead of the actual proxy, hence the extra ()  in `get_proxy_slot(n, tracer)()`.

It  also seems a little unfortunate that iiuc for tensor the slot holds an actual proxy still rather than a lambda.  Should we go all in on the lambda approach for consistency?  

or maybe better yet, we make set_proxy_slot smart enough to notice a 'deferred proxy' being set, and then get_proxy_slot automatically materializes the 'deferred proxy' before returning it to you?

Reply:
> It also seems a little unfortunate that iiuc for tensor the slot holds an actual proxy still rather than a lambda. Should we go all in on the lambda approach for consistency?

It actually doesn't; tensor stores a _ProxyTensor lol. Also, lazy proxies are not useful for tensors, where it is typically expected that every tensor computation actually gets used.

> or maybe better yet, we make set_proxy_slot smart enough to notice a 'deferred proxy' being set, and then get_proxy_slot automatically materializes the 'deferred proxy' before returning it to you?

@Chillee may have done this in his extension, let's see.